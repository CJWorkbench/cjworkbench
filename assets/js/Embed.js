import React from 'react'
import { escapeHtml, timeDifference } from './utils'
import { withI18n } from '@lingui/react'
import { t, Trans } from '@lingui/macro'
import PropTypes from 'prop-types'

export class Embed extends React.Component {
  static propTypes = {
    i18n: PropTypes.shape({
      // i18n object injected by LinguiJS withI18n()
      _: PropTypes.func.isRequired
    })
  }

  state = {
    overlayOpen: false
  }

  handleToggleOverlay = (e) => {
    this.setState({
      overlayOpen: !this.state.overlayOpen
    })
  }

  renderNotAvailable () {
    return (
      <div className='embed-wrapper'>
        <div className='embed-not-available'>
          <h1><Trans id='js.Embed.workflowNotAvailable.title'>This workflow is not available</Trans></h1>
        </div>
        <div className='embed-footer'>
          <div className='embed-footer-logo'>
            <img src={`${window.STATIC_URL}images/logo.png`} width='21' />
          </div>
          <div className='embed-footer-meta'>
            <h1>
              <Trans id='js.Embed.workflowNotAvailable.footer.logo' description='This should be all-caps for styling reasons'>
                    WORKBENCH
              </Trans>
            </h1>
          </div>
          <div className='embed-footer-button'>
            <i className='icon icon-share' />
          </div>
        </div>
      </div>
    )
  }

  render () {
    if (!this.props.workflow && !this.props.wf_module) {
      return this.renderNotAvailable()
    }
    const iframeCode = escapeHtml('<iframe src="' + window.location.protocol + '//' + window.location.host + '/embed/' + this.props.wf_module.id + '" width="560" height="315" frameborder="0" />')
    const timeAgo = timeDifference(this.props.workflow.last_update, new Date(), this.props.i18n)

    return (
      <div className='embed-wrapper'>
        <iframe src={'/api/wfmodules/' + this.props.wf_module.id + '/output'} frameBorder={0} />
        <div className='embed-footer'>
          <div className='metadata-stack'>
            <div className='embed-footer-logo'>
              <a href='http://workbenchdata.com' target='_blank' rel='noopener noreferrer'>
                <img src={`${window.STATIC_URL}images/logo.png`} width='35' />
              </a>
            </div>
            <div className='embed-footer-meta'>
              <div className='title'>
                <a href={'/workflows/' + this.props.workflow.id} target='_blank' rel='noopener noreferrer'>
                  {this.props.workflow.name}
                </a>
              </div>
              <div className='metadata'>
                <ul>
                  <li>
                    <a href={'/workflows/' + this.props.workflow.id} target='_blank' rel='noopener noreferrer'>
                    by {this.props.workflow.owner_name}
                    </a>
                  </li>
                  <li>
                    <a href={'/workflows/' + this.props.workflow.id} target='_blank' rel='noopener noreferrer'>
                      {this.props.i18n._(
                        /* i18n: {timeAgo} will contain a time difference (i.e. something like '4h ago') */
                        t('js.Embed.metadata.updated')`Updated ${timeAgo}`
                      )}
                    </a>
                  </li>
                </ul>
              </div>
            </div>
          </div>
          <button type='button' onClick={this.handleToggleOverlay} className='embed-footer-button'>
            <i className='icon icon-code' />
          </button>
        </div>
        <div className={'embed-overlay' + (this.state.overlayOpen ? ' open' : '')} onClick={this.handleToggleOverlay}>
          <div className='embed-share-links' onClick={(e) => { e.stopPropagation() }}>
            <h1><Trans id='js.Embed.embedThisChart' description='This should be all-caps for styling reasons'>EMBED THIS CHART</Trans></h1>
            <h2><Trans id='js.Embed.embedCode'>Paste this code into any webpage HTML</Trans></h2>
            <div className='code-snippet'>
              <code className='embed--share-code'>
                {iframeCode}
              </code>
            </div>
          </div>
        </div>
      </div>
    )
  }
}
export default withI18n()(Embed)
