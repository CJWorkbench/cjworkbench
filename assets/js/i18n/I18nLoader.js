import React from 'react'
import { I18nProvider } from '@lingui/react'
import fetchCatalog from './catalogs'

export class I18nLoader extends React.Component {
  state = {
    catalogs: {},
  }
  
  loadCatalog = (language) => {
      this.setState(state => ({
          catalogs: {
            ...state.catalogs,
            [language]: fetchCatalog(language)
          }
      }))
  }
  
  componentDidMount() {
    this.loadCatalog(this.props.language)
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { language } = nextProps
    const { catalogs } = nextState

    if (language !== this.props.language && !catalogs[language]) {
      this.loadCatalog(language)
      return false
    }

    return true
  }

  render () {
    const { children, language } = this.props
    const { catalogs } = this.state

    // Skip rendering when catalog isn't loaded.
    if (!catalogs[language]) {
        return null
    }

    return (
      <I18nProvider language={language} catalogs={catalogs}>
        {children}
      </I18nProvider>
    )
  }
}
