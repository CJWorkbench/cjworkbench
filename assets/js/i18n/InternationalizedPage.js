import React from 'react'
import PropTypes from 'prop-types'
import { en, el } from 'make-plural/plurals'
import { i18n } from '@lingui/core'
import { I18nProvider } from '@lingui/react'
import { currentLocaleId } from './locales'
import Spinner from '../Spinner'

__webpack_public_path__ = window.STATIC_URL + 'bundles/' // eslint-disable-line

async function loadLocale (localeId) {
  const { messages } = await import(
    /* webpackChunkName: "[request]" */
    `@lingui/loader!../../locale/${localeId}/messages.po`
  )
  i18n.loadLocaleData({
    en: { plurals: en },
    el: { plurals: el }
  })
  i18n.load(localeId, messages)
  i18n.activate(localeId)
  return i18n
}

function useActivatedI18n (localeId) {
  const [i18n, setI18n] = React.useState(null)

  React.useEffect(() => {
    loadLocale(currentLocaleId).then(setI18n)
  }, [])

  return i18n
}

export default function InternationalizedPage (props) {
  const { children } = props
  const i18n = useActivatedI18n(currentLocaleId)

  if (i18n === null) {
    return <Spinner />
  }

  return (
    <I18nProvider i18n={i18n} forceRenderOnLocaleChange={false}>
      {children}
    </I18nProvider>
  )
}
InternationalizedPage.propTypes = {
  children: PropTypes.node.isRequired
}
