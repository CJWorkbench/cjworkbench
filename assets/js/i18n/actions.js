const SET_LOCALE = 'SET_LOCALE';

function reduceSetLocale(state, action){
  return {
    ...state,
    locale: action.locale
  };
}

export function setLocaleAction(locale){
  return {
    type: SET_LOCALE,
    locale: locale
  }
}

export const reducerFunctions = {
  [SET_LOCALE]: reduceSetLocale
}

export function localeReducer(state, action){
  if (action.type in reducerFunctions) {
    // Run a registered reducer
    return reducerFunctions[action.type](state, action)
  }
  
  return state;
}
