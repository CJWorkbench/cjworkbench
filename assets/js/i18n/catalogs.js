import enCatalog from '../../locale/en/messages.po'
import elCatalog from '../../locale/el/messages.po'

const catalogs = {
  en: enCatalog,
  el: elCatalog
}

export default function fetchCatalog (language) {
  return catalogs[language]
}
