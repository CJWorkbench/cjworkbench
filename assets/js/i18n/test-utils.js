import { I18nProvider } from '@lingui/react'
import React from 'react'
import { shape, object } from 'prop-types'
import { mount, shallow } from 'enzyme'

/*
 * Below lie utils for testing with i18n.
 * See https://lingui.js.org/guides/testing.html
 */
// Create the I18nProvider to retrieve context for wrapping around.
const intlProvider = new I18nProvider({
  language: 'en',
  catalogs: {
    en: {}
  }
}, {})

const {
  linguiPublisher: {
    i18n: originalI18n
  }
} = intlProvider.getChildContext()

// You customize the i18n object here:
export const i18n = {
  ...originalI18n,
  // provide _ macro, for trying to unwrap default/id or just passing down the key
  _: key => {
    if (typeof key.defaults === 'string') return key.defaults
    if (typeof key.id === 'string') return key.id
    return key
  }
}

/**
 * When using Lingui `withI18n` on components, props.i18n is required.
 */
function nodeWithI18nProp (node) {
  return React.cloneElement(node, { i18n })
}

/*
 * Methods to use
 */
export function shallowWithI18n (node, { context } = {}) {
  return shallow(
    nodeWithI18nProp(node),
    {
      context: Object.assign({}, context, { i18n })
    }
  )
}

export function mountWithI18n (node, { context, childContextTypes } = {}) {
  const newContext = Object.assign({}, context, { linguiPublisher: { i18n } })
  /*
   * I18nProvider sets the linguiPublisher in the context for withI18n to get
   * the i18n object from.
   */
  const newChildContextTypes = Object.assign({},
    {
      linguiPublisher: shape({
        i18n: object.isRequired
      }).isRequired
    },
    childContextTypes
  )
  return mount(
    nodeWithI18nProp(node),
    {
      context: newContext,
      childContextTypes: newChildContextTypes
    }
  )
}
