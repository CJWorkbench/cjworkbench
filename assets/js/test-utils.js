/* globals jest */
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import promiseMiddleware from 'redux-promise-middleware'
import errorMiddleware from './error-middleware'
import { workflowReducer } from './workflow-reducer'

// Returns new mock function that returns given json. Used for mocking "get" API calls
export function jsonResponseMock (json) {
  return jest.fn().mockImplementation(() =>
    Promise.resolve(json)
  )
}

// Returns new mock function that gives an OK HTTP response. Use for mocking "set" API calls
export function okResponseMock () {
  return jsonResponseMock(null)
}

// Helper function to return a promise that resolves after all other promise mocks,
// even if they are chained like Promise.resolve().then(...).then(...)
// Technically: this is designed to resolve on the next macrotask
// https://stackoverflow.com/questions/25915634/difference-between-microtask-and-macrotask-within-an-event-loop-context
export function tick () {
  return new Promise(resolve => {
    setTimeout(resolve, 0)
  })
}

export function sleep (ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}

/**
 * Return `[ notify, promise ]`; `promise` completes when `notify` is called.
 *
 * Usage:
 *
 *     const [ notify, condition ] = createConditionVariable
 *     const done = condition.then(do_something)
 *     await Promise.all([sleep(0.1).then(notify), done])
 */
export function createConditionVariable () {
  let notify
  const promise = new Promise(resolve => { notify = resolve })
  return [notify, promise]
}

/**
 * Mock Redux store with in-production reducers.
 *
 * This is for integration-style tests: they test the reducer _and_ the caller
 * at the same time.
 *
 * Usage:
 *
 *     const api = { undo: jest.fn() }
 *     const store = mockStoreWithReducer({'workflow': ..., ...}, api)
 *     store.dispatch(Actions.undo())
 *     expect(api.undo).toHaveBeenCalled()
 *     expect(store.getState().workflow).toEqual(...)
 */
export function mockStore (initialState, api = {}) {
  const middlewares = [errorMiddleware(), promiseMiddleware, thunk.withExtraArgument(api)]
  const store = createStore(workflowReducer, initialState, applyMiddleware(...middlewares))
  return store
}
