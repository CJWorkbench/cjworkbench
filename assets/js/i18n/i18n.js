import { setupI18n } from "@lingui/core"
import { currentLocale } from './locales'
import { fetchCatalog } from './catalogs'

export default const i18n = setupI18n({
   language: currentLocale,
   catalogs: {
       currentLocale: fetchCatalog(currentLocale)
   }
})
