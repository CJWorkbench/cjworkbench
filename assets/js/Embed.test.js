/* globals beforeEach, describe, expect, it */
import React from 'react'
import { Embed } from './Embed'
import { shallowWithI18n } from './i18n/test-utils'

describe('Embed', () => {
  let wrapper

  describe('Available workflow', () => {
    beforeEach(() => {
      wrapper = shallowWithI18n(
        <Embed
          wf_module={{
            id: 1
          }}
          workflow={{
            name: 'Workflow Title',
            owner_name: 'Workflow Owner Name',
            id: 1
          }}
        />
      )
    })

    it('Renders the embed widget with the correct information', () => {
      expect(wrapper).toMatchSnapshot()
      expect(wrapper.find('.embed-footer-meta .title').text()).toBe('Workflow Title')
    })

    it('Displays the sharing overlay', () => {
      wrapper.find('.embed-footer-button').simulate('click')
      expect(wrapper.find('.embed-overlay').hasClass('open')).toBe(true)
    })
  })

  describe('Unavailable workflow', () => {
    beforeEach(() => {
      wrapper = shallowWithI18n(
        <Embed
          wf_module={null}
          workflow={null}
        />
      )
    })

    it('Renders the embed widget with the correct information', () => {
      expect(wrapper).toMatchSnapshot()
      expect(wrapper.find('.embed-not-available').length).toBe(1)
    })
  })
})
