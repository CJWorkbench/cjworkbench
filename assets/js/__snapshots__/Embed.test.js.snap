// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Embed Available workflow Renders the embed widget with the correct information 1`] = `
<div
  className="embed-wrapper"
>
  <iframe
    frameBorder={0}
    src="/api/wfmodules/1/output"
  />
  <div
    className="embed-footer"
  >
    <div
      className="metadata-stack"
    >
      <div
        className="embed-footer-logo"
      >
        <a
          href="http://workbenchdata.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <img
            src="undefinedimages/logo.png"
            width="35"
          />
        </a>
      </div>
      <div
        className="embed-footer-meta"
      >
        <div
          className="title"
        >
          <a
            href="/workflows/1"
            rel="noopener noreferrer"
            target="_blank"
          >
            Workflow Title
          </a>
        </div>
        <div
          className="metadata"
        >
          <ul>
            <li>
              <a
                href="/workflows/1"
                rel="noopener noreferrer"
                target="_blank"
              >
                by 
                Workflow Owner Name
              </a>
            </li>
            <li>
              <a
                href="/workflows/1"
                rel="noopener noreferrer"
                target="_blank"
              >
                Updated {timeAgo}
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <button
      className="embed-footer-button"
      onClick={[Function]}
      type="button"
    >
      <i
        className="icon icon-code"
      />
    </button>
  </div>
  <div
    className="embed-overlay"
    onClick={[Function]}
  >
    <div
      className="embed-share-links"
      onClick={[Function]}
    >
      <h1>
        <WithI18n
          defaults="EMBED THIS CHART"
          description="This should be all-caps for styling reasons"
          id="js.Embed.embedThisChart"
        />
      </h1>
      <h2>
        <WithI18n
          defaults="Paste this code into any webpage HTML"
          id="js.Embed.embedCode"
        />
      </h2>
      <div
        className="code-snippet"
      >
        <code
          className="embed--share-code"
        >
          &lt;iframe src="http://localhost/embed/1" width="560" height="315" frameborder="0" /&gt;
        </code>
      </div>
    </div>
  </div>
</div>
`;

exports[`Embed Unavailable workflow Renders the embed widget with the correct information 1`] = `
<div
  className="embed-wrapper"
>
  <div
    className="embed-not-available"
  >
    <h1>
      <WithI18n
        defaults="This workflow is not available"
        id="js.Embed.workflowNotAvailable.title"
      />
    </h1>
  </div>
  <div
    className="embed-footer"
  >
    <div
      className="embed-footer-logo"
    >
      <img
        src="undefinedimages/logo.png"
        width="21"
      />
    </div>
    <div
      className="embed-footer-meta"
    >
      <h1>
        <WithI18n
          defaults="WORKBENCH"
          description="This should be all-caps for styling reasons"
          id="js.Embed.workflowNotAvailable.footer.logo"
        />
      </h1>
    </div>
    <div
      className="embed-footer-button"
    >
      <i
        className="icon icon-share"
      />
    </div>
  </div>
</div>
`;
